test/sslapitest.c

```diff
@@ -1188,91 +1210,88 @@ static int test_ktls_sendfile(int tls_version, const char *cipher)
 }
 
 #if !defined(OPENSSL_NO_TLS1_2) || !defined(OPENSSL_NO_TLS1_3)
+static struct ktls_test_cipher {
+    int tls_version;
+    const char *cipher;
+} ktls_test_ciphers[] = {
+# if !defined(OPENSSL_NO_TLS1_2)
+#  ifdef OPENSSL_KTLS_AES_GCM_128
+    { TLS1_2_VERSION, "AES128-GCM-SHA256" },
+#  endif
+#  ifdef OPENSSL_KTLS_AES_CCM_128
+    { TLS1_2_VERSION, "AES128-CCM"},
+#  endif
+#  ifdef OPENSSL_KTLS_AES_GCM_256
+    { TLS1_2_VERSION, "AES256-GCM-SHA384"},
+#  endif
+#  ifdef OPENSSL_KTLS_CHACHA20
+    { TLS1_2_VERSION, "ECDHE-RSA-CHACHA20-POLY1305"},
+#  endif
+#endif
+#if !defined(OPENSSL_NO_TLS1_3)
+#  ifdef OPENSSL_KTLS_AES_GCM_128
+    { TLS1_3_VERSION, "TLS_AES_128_GCM_SHA256" },
+#  endif
+#  ifdef OPENSSL_KTLS_AES_CCM_128
+    { TLS1_3_VERSION, "TLS_AES_128_CCM_SHA256" },
+#  endif
+#  ifdef OPENSSL_KTLS_AES_GCM_256
+    { TLS1_3_VERSION, "TLS_AES_256_GCM_SHA384" },
+#  endif
+#  ifdef OPENSSL_KTLS_CHACHA20
+    { TLS1_3_VERSION, "TLS_CHACHA20_POLY1305_SHA256" },
+#  endif
+#endif
+};
+
+#define NUM_KTLS_TEST_CIPHERS \
+    (sizeof(ktls_test_ciphers) / sizeof(ktls_test_ciphers[0]))
+
 static int test_ktls(int test)
 {
+    struct ktls_test_cipher *cipher;
     int cis_ktls_tx, cis_ktls_rx, sis_ktls_tx, sis_ktls_rx;
-    int tlsver, testresult;
 
-    if (test > 15) {
-#if defined(OPENSSL_NO_TLS1_3)
-        return 1;
-#else
-        test -= 16;
-        tlsver = TLS1_3_VERSION;
-#endif
-    } else {
-#if defined(OPENSSL_NO_TLS1_2)
-        return 1;
-#else
-        tlsver = TLS1_2_VERSION;
-#endif
-    }
+    OPENSSL_assert(test / 16 < NUM_KTLS_TEST_CIPHERS);
+    cipher = &ktls_test_ciphers[test / 16];
 
     cis_ktls_tx = (test & 1) != 0;
     cis_ktls_rx = (test & 2) != 0;
     sis_ktls_tx = (test & 4) != 0;
     sis_ktls_rx = (test & 8) != 0;
 
+#ifdef OPENSSL_KTLS_CHACHA20
+    /* Skip ChaCha20Poly1305 as currently FIPS module does not support it */
+    if (strstr(cipher->cipher, "CHACHA20") != NULL && is_fips == 1)
```

Comment:
Hmmm. This is strange. I guess this is needed because we pre-fetch the `EVP_CIPHER` even though we never actually end up using it because we use the kernel implementation instead. It seems odd that we have to have a provider implementation available  even though we never actually use it...

...OTOH, KTLS usage when we're expecting to be using the FIPS provider is a bit of a concern. Essentially we would end up using non-validated crypto in this case!!

Reply:
Well in theory (and in practice as well with RHEL for example) you can have FIPS validated kernel crypto modules.

Reply:
> Well in theory (and in practice as well with RHEL for example) you can have FIPS validated kernel crypto modules.

Ok - true - but I reckon in many cases people will hit this by accident - thinking that they are using the OpenSSL FIPS module when they're not.

Reply:
Yeah, for that reason I think KTLS support should be run-time optional and disabled by default. There should be a configuration file option to enable it. The current way when KTLS is unconditionally used once built-in and the supported algorithms match seems wrong to me, not just because of FIPS.

Reply:
> Yeah, for that reason I think KTLS support should be run-time optional and disabled by default.

That seems reasonable. At the moment it is not built by default, but once built it is always on at runtime. We could switch that around so that it is always built if possible, but it is disabled at runtime unless you explicitly ask for it.

Reply:
I raised this as a separate issue (#13794), so as not to derail this PR with this tangential problem.

Reply:
I'm fine with having an option to control KTLS, but I do think that such a change is probably orthogonal to this PR and belongs in a separate PR.  Also, it sounds like for the KTLS tests we'd then have to have a separate config file added to the mix at least for running the test_sslapi?  And do you have any guidance as to which layer of config a KTLS knob would need to be added?  (The EtM setting for CBC is the only config option I've used, and I can use that as a model if you think that is at the right section, etc.)  Presumably the config knob would need to have separate knobs for both RX and TX given the existing API extensions for KTLS support disabling KTLS for both RX and TX independently.

Reply:
Well, ordinarily, we would have added KTLS as an "option": e.g. SSL_OP_ENABLE_KTLS. This would mean we could use the existing ssl_conf.c code for handling options (see the function `cmd_Options` in that file). However we seem to have decided to add KTLS as a "mode" instead (I don't recall why we did that - but my guess is because the SSL_OP_* space is almost out of a spare "bits" to add more options). I don't think there is any code in ssl_conf.c for configuring modes - but if we decided to add some I guess it would look much llike the "Options" code.